### PR TITLE
savepoint

### PR DESCRIPTION
**Pull request by borkarsaish65**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: module/library/categories/helper.js [hunk 1]
```diff
@@ -321,9 +394,65 @@ module.exports = class LibraryCategoriesHelper {
 	 * @returns {Object} category details
 	 */
 
-	static create(categoryData) {
+	static create(categoryData, files) {
```
Comment (VISHNUDAS-tunerlabs) [inline module/library/categories/helper.js:397]: @borkarsaish65 Please update function signature

File: module/library/categories/helper.js [hunk 2]
```diff
@@ -321,9 +394,65 @@ module.exports = class LibraryCategoriesHelper {
...
 		return new Promise(async (resolve, reject) => {
 			try {
+				if (files && files.cover_image) {
+					let coverImages = files.cover_image
```
Comment (VISHNUDAS-tunerlabs) [inline module/library/categories/helper.js:401]: @borkarsaish65 we support passing multiple files during category creation right!

File: module/library/categories/helper.js [hunk 3]
```diff
@@ -321,9 +394,65 @@ module.exports = class LibraryCategoriesHelper {
...
+								},
+							})
+
+							if (!(uploadData.status == 200 || uploadData.status == 201)) {
+								throw new Error('File upload failed.')
+							}
+						}
+					}
+
+					let i = 0
```
Comment (VISHNUDAS-tunerlabs) [inline module/library/categories/helper.js:445]: @borkarsaish65 instead of i call it sequenceNumber

File: module/library/categories/helper.js [hunk 4]
```diff
@@ -321,9 +394,65 @@ module.exports = class LibraryCategoriesHelper {
...
+							const uploadData = await axios.put(fileUploadUrl[0].url, fileFromRequest.data, {
+								headers: {
+									'x-ms-blob-type':
+										process.env.CLOUD_STORAGE_PROVIDER === 'azure' ? 'BlockBlob' : null,
+									'Content-Type': 'multipart/form-data',
+								},
+							})
+
+							if (!(uploadData.status == 200 || uploadData.status == 201)) {
+								throw new Error('File upload failed.')
```
Comment (VISHNUDAS-tunerlabs) [inline module/library/categories/helper.js:440]: @borkarsaish65 always throw error with proper status and message

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed NOv 29 2:13 PM

File: module/library/categories/helper.js [hunk 5]
```diff
@@ -173,6 +174,78 @@ module.exports = class LibraryCategoriesHelper {
 					})
 				}
 
+				let projectTemplates = result[0].data
+				let filePathsArray = []
+				for (let project of projectTemplates) {
+					let categories = project.categories
+
+					let categoryIdArray = categories.map((category) => {
```
Comment (VISHNUDAS-tunerlabs) [inline module/library/categories/helper.js:182]: @borkarsaish65 handle situation if _id is not present

File: module/library/categories/helper.js [hunk 6]
```diff
@@ -173,6 +174,78 @@ module.exports = class LibraryCategoriesHelper {
 					})
 				}
 
+				let projectTemplates = result[0].data
+				let filePathsArray = []
+				for (let project of projectTemplates) {
```
Comment (VISHNUDAS-tunerlabs) [inline module/library/categories/helper.js:179]: @borkarsaish65 can we try this
if front end requet 10 project-templates
fetch them
collect all the category ids from all ten templates
make the array unique
get evidences of categories using unique array of ids
find downloadable url for them in a single call
map back the evidence details to project template data based on the category ids present in them

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]

File: module/library/categories/helper.js (hunk 1 above)
Comment (borkarsaish65) [reply module/library/categories/helper.js:397]: done

File: module/library/categories/helper.js (hunk 2 above)
Comment (borkarsaish65) [reply module/library/categories/helper.js:401]: yes

File: module/library/categories/helper.js (hunk 3 above)
Comment (borkarsaish65) [reply module/library/categories/helper.js:445]: done

File: module/library/categories/helper.js (hunk 6 above)
Comment (borkarsaish65) [reply module/library/categories/helper.js:181]: changed the logic now

File: module/library/categories/helper.js (hunk 5 above)
Comment (borkarsaish65) [reply module/library/categories/helper.js]: done

File: module/library/categories/helper.js (hunk 4 above)
Comment (borkarsaish65) [reply module/library/categories/helper.js]: done

File: module/library/categories/helper.js [hunk 7]
```diff
@@ -173,6 +174,94 @@ module.exports = class LibraryCategoriesHelper {
...
+							projectCategory.evidences = singleCategoryInfo.evidences
+						}
+					}
+				}
+
+				let allFilePaths = filePathsArray.map((project) => {
+					return project.filePaths
+				})
+
+				let finalArr = _.flatten(allFilePaths)
```
Comment (VISHNUDAS-tunerlabs) [inline module/library/categories/helper.js:229]: @borkarsaish65 please rename the variable to more meaning full word, Please add some comments

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]

**Review by VISHNUDAS-tunerlabs — APPROVED** [review]

**Pull request merged**